orttraining/orttraining/training_ops/cuda/reduction/reduction_ops.cc

```diff
@@ -59,7 +59,7 @@ Status ReduceKernel<allow_multi_axes>::ComputeImplEx(OpKernelContext* ctx, cudnn
   Tensor* Y = ctx->Output(0, prepare_reduce_metadata.squeezed_output_dims);
   bool fast_reduction = fast_reduction_;
   if (fast_reduction) {
-    auto ctx_internal = static_cast<OpKernelContextInternal*>(ctx);
+    auto ctx_internal = dynamic_cast<OpKernelContextInternal*>(ctx);
     if (ctx_internal && ctx_internal->GetUseDeterministicCompute())
```

Comment:
Probably a dumb question. But why do we need `dynamic` casting?

Reply:
I noticed the check for ctx_internal (!= nullptr). from that, i assumed that the intent of the original code was to downcast and check if it was successful. static_cast doesn't do a runtime check and return nullptr on failure, but dynamic_cast does.
[https://en.cppreference.com/w/cpp/language/static_cast](https://en.cppreference.com/w/cpp/language/static_cast)
[https://en.cppreference.com/w/cpp/language/dynamic_cast](https://en.cppreference.com/w/cpp/language/dynamic_cast)


---
In reply to: [517614785](https://github.com/microsoft/onnxruntime/pull/5639#discussion_r517614785) [](ancestors = 517614785)

Reply:
I feel we should throw if the casting fails. It fails only when ORT wrongly enters this function.

---
In reply to: [517626380](https://github.com/microsoft/onnxruntime/pull/5639#discussion_r517626380) [](ancestors = 517626380,517614785)

Reply:
is it the case that it is invalid for ctx to be anything but OpKernelContextInternal* here?

---
In reply to: [517749675](https://github.com/microsoft/onnxruntime/pull/5639#discussion_r517749675) [](ancestors = 517749675,517626380,517614785)

Reply:
If `GetUseDeterministicCompue` is a member of `OpKernelContextInternal`, then `ctx` must be `Op...`.

---
In reply to: [517754476](https://github.com/microsoft/onnxruntime/pull/5639#discussion_r517754476) [](ancestors = 517754476,517749675,517626380,517614785)

Reply:
oh, what i meant by "here" is within this function instead of within this if block. i'm not sure we can assume that OpKernelContext* ctx should actually point to an OpKernelContextInternal instance.

---
In reply to: [517767805](https://github.com/microsoft/onnxruntime/pull/5639#discussion_r517767805) [](ancestors = 517767805,517754476,517749675,517626380,517614785)

Reply:
You're right. I am not sure too, so it's another topic not related to this PR. You can merge this PR and discuss this in Core meeting. :)

---
In reply to: [517777926](https://github.com/microsoft/onnxruntime/pull/5639#discussion_r517777926) [](ancestors = 517777926,517767805,517754476,517749675,517626380,517614785)